Sources/Apexy_URLSession/URLSessionClient.swift

```diff
@@ -0,0 +1,189 @@
+import Foundation
+import Apexy
+
+final public class URLSessionClient: Client {
+    
+    /// A closure used to observe result of every response from the server.
+    public typealias ResponseObserver = (URLRequest?, HTTPURLResponse?, Data?, Error?) -> Void
+
+    private let session: URLSession
+    
+    private let requestAdapter: RequestAdapter
+
+    /// The queue on which the completion handler is dispatched.
+    private let completionQueue: DispatchQueue
+
+    /// This closure to be called after each response from the server for the request.
+    private let responseObserver: ResponseObserver?
+
+    private let sessionDelegate = SessionDelegate()
+    
+    /// Creates new 'URLSessionClient' instance.
+    ///
+    /// - Parameters:
+    ///   - baseURL: Base `URL`.
+    ///   - configuration: The configuration used to construct the managed session.
+    ///   - completionQueue: The serial operation queue used to dispatch all completion handlers. `.main` by default.
+    ///   - responseObserver: The closure to be called after each response.
+    public init(
+        baseURL: URL,
+        configuration: URLSessionConfiguration = .default,
+        completionQueue: DispatchQueue = .main,
+        responseObserver: ResponseObserver? = nil) {
+        
+        self.requestAdapter = BaseRequestAdapter(baseURL: baseURL)
+        self.session = URLSession(configuration: configuration, delegate: sessionDelegate, delegateQueue: .main)
+        self.completionQueue = completionQueue
+        self.responseObserver = responseObserver
+    }
+    
+    public func request<T>(
+        _ endpoint: T,
+        completionHandler: @escaping (APIResult<T.Content>) -> Void) -> Progress where T : Endpoint {
+        
+        var request: URLRequest
+        do {
+            request = try endpoint.makeRequest()
+            request = try requestAdapter.adapt(request).get()
+        } catch {
+            completionHandler(.failure(error))
+            return Progress()
+        }
+        
+        let task = session.dataTask(with: request) { (data, response, error) in
```

Comment:
URLSession closure based api and delegate api  are mutually exclusive.

You only need to use the dalaget api